File: 1285-balance-a-binary-search-tree/balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void pushInSortedOrder(TreeNode* root, vector<int>& sorted_values){
        if(root == NULL) return;

        pushInSortedOrder(root->left, sorted_values);
        sorted_values.push_back(root->val);
        pushInSortedOrder(root->right, sorted_values);
    }

    TreeNode* splitSorted(vector<int>& sorted_values, int low, int high){
        if(low > high) return NULL;

        int mid = (low + high)/2;
        TreeNode* newRoot = new TreeNode(sorted_values[mid]);
        newRoot->left = splitSorted(sorted_values, low, mid-1);
        newRoot->right = splitSorted(sorted_values, mid+1, high);
        return newRoot;
    }

    TreeNode* balanceBST(TreeNode* root) {
        vector<int> sorted_values = {};
        pushInSortedOrder(root, sorted_values);
        TreeNode* bst = splitSorted(sorted_values, 0, sorted_values.size() - 1);
        return bst;
    }


    

    
};

File: 1285-balance-a-binary-search-tree/balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    # So we just get all the values then make a new tree thats balanced.
    # If all the values are sorted: middle element is root, then balance based on the two subtrees
    # Since its given as BST, we can get sorted then create in O(N)

    def fillSortedArr(self, root):
        if root == None:
            return
        self.fillSortedArr(root.left)
        self.sortedArr.append(root.val)
        self.fillSortedArr(root.right)
    
    def createBalancedTree(self, startInd, endInd):
        midInd = (startInd + endInd) // 2
        rootNode = TreeNode(self.sortedArr[midInd])
        if startInd < midInd:
            rootNode.left = self.createBalancedTree(startInd, midInd - 1)
        
        if endInd > midInd:
            rootNode.right = self.createBalancedTree(midInd + 1, endInd)
        
        return rootNode

    def balanceBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.sortedArr = []
        self.fillSortedArr(root)
        return self.createBalancedTree(0, len(self.sortedArr) - 1)